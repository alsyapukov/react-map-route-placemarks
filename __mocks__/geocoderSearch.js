import axios from "./axios";

// Наши настройки api
export const apiSettings = {
  url: 'http://geocode-maps.yandex.ru/1.x/',
  params: {
    apikey: 'c0bc84b6-7dec-45b4-be8a-85945e742d4b',
    version: '1.x',
    results: '1',
    format: 'json'
  }
}

// Делаем запрос в геокодер Яндекса, на тесте можно ставить и тестовый ключ.
// 25000 запросов в день можно слать бесплатно
export default async term => {
  const response = await axios.get(apiSettings.url, {
    params: {
      ...apiSettings.params,
      geocode: term
    }
  });

  return response;
};