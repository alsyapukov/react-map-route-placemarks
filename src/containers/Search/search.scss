.search {
  width: 100%;

  &__textbox {
    position: sticky;
    top: 0;
    z-index: 2;
    background-color: #fff;
    padding-bottom: 20px;
  }
  &__list {
    max-height: 100%;
    position: relative;
    overflow: auto;
  }
}
.balloon {
  display: inline-block;
  padding: 20px;
  min-height: 75px;
  position: relative;
  bottom: 80px;
  left: 0;
  width: 200px;
  max-width: 250px;
  color: #333333;
  background-color: #fff;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.2);
  .baloon__close {
    margin-top: -3px;
  }
}