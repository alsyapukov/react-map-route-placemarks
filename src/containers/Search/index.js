import React, { Component } from "react";
import './search.scss'
import TextBox from '../../components/base/TextBox'

import { Search as SearchGeocode } from '../../api'

import SortSearch from '../SortSearch'

class Search extends Component {

  state = {
    value: "",
    searchArray: [],
    polyline: null,
    placemarks: []
  }

  changeValue = (val) => {
    this.setState({ value: val });
  }

  geocoderSearch = (val) => {
    SearchGeocode.getAddress({ geocode: val })
      .then(res => {
        const geoObject = res.data.response.GeoObjectCollection.featureMember[0];
        if (geoObject) {

          let searchValue = {
            address: geoObject.GeoObject.metaDataProperty.GeocoderMetaData.text,
            coords: geoObject.GeoObject.Point.pos.split(' ').map(coord => {
              return Number(coord);
            }).reverse()
          }

          this.setState({
            searchArray: [...this.state.searchArray, searchValue]
          })
          
          this.initPlacemark(searchValue, this.state.searchArray.length - 1);
          this.setCenter();
        }
      })
  }

  handleKeyPress = async(e) => {
    if (e.key === 'Enter') {
      this.geocoderSearch(this.state.value);
    }
  }

  removeSearchItem = (i) => {
    let newArray = this.state.searchArray;

    newArray = newArray.filter((item, index) => index !== i);
    this.setState({
      searchArray: newArray
    })
    this.props.myMap.geoObjects.remove(this.state.placemarks[i]);

    let placemarks = [];
    placemarks = [...placemarks, ...this.state.placemarks];
    placemarks.splice(i, 1);

    this.setState({
      placemarks: placemarks
    })
    if (newArray.length) {
      this.changePolyline(newArray)
    }
  }

  setCenter = () => {
    this.props.myMap.setCenter(
      this.state.searchArray[this.state.searchArray.length - 1].coords
    );
  }

  dragEventMarks = () => {
    this.props.myMap.geoObjects.events.add(['dragend'], (e) => {

      let placemark = e.get('target');
      let coords = e.get('target').geometry.getCoordinates();
      let index = placemark.properties.getAll().index;

      ymaps
        .geocode(coords, {
          results: 1
        })
        .then(res => {
          let firstGeoObject = res.geoObjects.get(0);
          let searchValue = {
            address: firstGeoObject.getAddressLine(),
            coords
          }

          placemark.properties.set('balloonContent', searchValue.address);

          let newArray = this.state.searchArray
            .map((item, i) => index == i ? searchValue : item)
          this.changePolyline(newArray);
          this.changePlacemark(newArray);
          this.setState({
            searchArray: newArray
          })
        });

    });
  }

  initPlacemark = (val, index) => {
    var placemark = new ymaps.Placemark(
      val.coords,
      {
        balloonContent: `${val.address}`,
        index: index
      },
      {
        preset: 'islands#circleDotIcon',
        iconColor: 'blue',
        draggable: true
      }
    );
    this.setState({
      placemarks: [...this.state.placemarks, placemark]
    })

    if (this.state.searchArray.length == 1) {
      this.dragEventMarks();
      this.initPolyline(this.state.searchArray);
      this.changePolyline(this.state.searchArray);
    } else if (this.state.searchArray.length > 1) {
      this.changePolyline(this.state.searchArray);
    }

    this.props.myMap.geoObjects.add(placemark);
  }

  changePlacemark = (newArray) => {
    newArray.map((item, i) => {
      this.state.placemarks[i].geometry.setCoordinates(item.coords);
    })
  }

  initPolyline = (val) => {
    val = val.map(item => {
      return item.coords
    })
    this.setState({
      polyline: new ymaps.Polyline(val, {
        balloonContent: "Route between coordinates"
      }, {
        strokeColor: "#000000",
        strokeWidth: 10,
        strokeOpacity: 0.4
      })
    })
    this.props.myMap.geoObjects.add(this.state.polyline);
  }

  changePolyline = (val) => {
    val = val.map(item => {
      return item.coords
    })
    this.state.polyline.geometry.setCoordinates(val)
  }

  sort = (val) => {
    this.setState({
      searchArray: val
    })
    this.changePlacemark(val);
    this.changePolyline(val);
  }

  render() {
    return (
      <div className="search" >
        <div className="search__textbox">
          <TextBox changeValue={this.changeValue} onKeyPress={this.handleKeyPress} />
        </div>
        <div className="search__list">
          <SortSearch items={this.state.searchArray} sort={(e) => this.sort(e)} removeSearchItem={(e) => this.removeSearchItem(e)} />
        </div>
      </div>
    );
  }
}
export default Search;