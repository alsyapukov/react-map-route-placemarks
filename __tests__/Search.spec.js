import React from 'react';
import { create } from 'react-test-renderer';
import Search from '../src/containers/Search'

import mockAxios from "../__mocks__/axios";
import geocoderSearch, { apiSettings } from "../__mocks__/geocoderSearch";


describe('Test Search component', () => {

  it('renders Search component', () => {
    const component = create(<Search />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("Check geocoderSearch function", async () => {

    // Макет нашей функции
    mockAxios.get.mockImplementationOnce(() =>
      Promise.resolve([
        {
          address: "Россия, Москва, Красная площадь",
          coords: [55.753595, 37.621031]
        }
      ])
    );

    // Вызываем нашу функцию с поисковой строкой
    const searchFunc = await geocoderSearch('Красная площадь');

    // Ожидаемое значение
    const expectSearch = [
      {
        address: "Россия, Москва, Красная площадь",
        coords: [55.753595, 37.621031]
      }
    ];

    // сравниваем значение из функции и ожидаемое значение
    expect(expect.arrayContaining(searchFunc)).toEqual(expect.arrayContaining(expectSearch));

    // Проверили что она вызывается 1 раз
    expect(mockAxios.get).toHaveBeenCalledTimes(1);
    
    // Проверяем вызвана ли функция с конкретными аргументами
    expect(mockAxios.get).toHaveBeenCalledWith(
      apiSettings.url,
      {
        params: {
          ...apiSettings.params,
          geocode: 'Красная площадь'
        }
      }
    );
  });
});