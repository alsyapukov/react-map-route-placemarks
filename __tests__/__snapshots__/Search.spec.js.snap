// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Search component renders Search component 1`] = `
<div
  className="search"
>
  <div
    className="search__textbox"
  >
    <div
      className="textbox"
    >
      <input
        className="textbox__inner"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="search__list"
  >
    <ul />
  </div>
</div>
`;
